docs/design/proposals/rbd-volume-healer.md

```diff
@@ -0,0 +1,102 @@
+# RBD NBD VOLUME HEALER
+
+- [RBD NBD VOLUME HEALER](#rbd-nbd-volume-healer)
+  - [Rbd Nbd](#rbd-nbd)
+    - [Advantages of userspace mounters](#advantages-of-userspace-mounters)
+    - [Side effects of userspace mounters](#side-effects-of-userspace-mounters)
+  - [Volume Healer](#volume-healer)
+    - [More thoughts](#more-thoughts)
+
+## Rbd nbd
+
+The rbd CSI plugin will provision new rbd images and attach and mount those
+to workloads. Currently, the default mounter is krbd, which uses the kernel
+rbd driver to mount the rbd images onto the application pod. Here on
+at Ceph-CSI we will also have a userspace way of mounting the rbd images,
+via rbd-nbd.
+
+[Rbd-nbd](https://docs.ceph.com/en/latest/man/8/rbd-nbd/) is a client for
+RADOS block device (rbd) images like the existing rbd kernel module. It
+will map an rbd image to an nbd (Network Block Device) device, allowing
+access to it as a regular local block device.
+
+![csi-rbd-nbd](./images/csi-rbd-nbd.svg)
+
+It’s worth making a note that the rbd-nbd processes will run on the
+client-side, which is inside the `csi-rbdplugin` node plugin.
+
+### Advantages of userspace mounters
+
+- It is easier to add features to rbd-nbd as it is released regularly with
+  Ceph, and more difficult and time consuming to add features to the kernel
+  rbd module as that is part of the Linux kernel release schedule.
+- Container upgrades will be independent of the host node, which means if
+  there are any new features with rbd-nbd, we don’t have to reboot the node
+  as the changes will be shipped inside the container.
+- Because the container upgrades are host node independent, we will be a
+  better citizen in K8s by switching to the userspace model.
+- Unlike krbd, rbd-nbd uses librbd user-space library that gets most of the
+  development focus, and hence rbd-nbd will be feature-rich.
+- Unlike rbd-nbd, krbd exhibits additional drawback, being entirely kernel
+  space impacts fault-tolerance as any kernel panic affects a whole node
+  not only a single pod that is using rbd storage. Thanks to the rbd-nbd’s
```

Comment:
considering rbd nbd running in the node, isnt this equally applicable for rbd nbd as well ?

Reply:
There is a difference, the krbd is a complete kernel and vendor-specific driver which needs changes on every feature basis. On the other hand, NBD is a generic driver, while all the vendor-specific logic sits in the userspace.

Reply:
I should have been more clear, I was pointing out `kernel panic` situation.

Reply:
Do you see this as a better fit?

```
Being entirely kernel space impacts fault-tolerance as any kernel panic  
affects a whole node not only a single pod that is using rbd storage.    
Thanks to the rbd-nbd’s userspace design, we are less bothered here, the 
krbd is a complete kernel and vendor-specific driver which needs changes 
on every feature basis, on the other hand, rbd-nbd depends on NBD generic
driver, while all the vendor-specific logic sits in the userspace. It's  
worth taking note that NBD generic driver is mostly unchanged much from  
years and consider it to be much stable. Also given NBD is a generic     
driver there will be many eyes on it compared to the rbd driver.         
```

Reply:
@pkalever not sure, the only thing which I was pointing here is any `panic` situation in the kernel where mount happened equally applies the krbd or nbd as long as 'nbd` runs inside the pod, does that make sense ? so I am not sure we have to specifically mention this `panic` case rather start from here directly .. ie ``` krbd is a complete kernel and vendor-specific driver ....` , wdyt?

Reply:
The intention was to highlight the fact that rbd-nbd gets IO logic in userspace while krbd has to handle them in kernel completely. And there is an advantage for carrying the vendor-specific logic in the userspace rather than keeping it in the kernel space, frequently changing logic might be more prone to bugs which might lead to panics.

Reply:
@pkalever yeah, I get that, so this `krbd is a complete kernel and vendor specific ..` line is actually conveying that message.  Any way, please let me know if you would like just adjust above or go ahead with current version.  I am fine with any.

Reply:
Thanks @humblec, I have adjusted the previous version with the version mentioned in this comment.

Reply:
Thanks a lot @pkalever .. looks good :+1: